### DISCUSSION TITLE
Scale horizontally using Redis

### DISCUSSION THREAD

**Discussion by danieleratti** [uNetworking/uWebSockets#954] [Q&A] [answered]
Is there any possibility of scaling uWebSockets on multiple instances (for eg. using Redis) as it had been possible in past with Socket.io + the old uws using the io.adapter method?

`io.adapter(redis({ ... }));`

**Comment by ghost** [comment, marked as answer]
It's not more complicated than publishing to Redis every time one of your instances end up publishing to their clients. Then the other instances connected to Redis will also publish to their clients.

You don't even need Redis for this, you just need some client and some server to act as "master". I'm pretty sure Kubernetes and the like already has support for master/slave set-ups like this.

**Comment by vahansoft** [comment]
> It's not more complicated than publishing to Redis every time one of your instances end up publishing to their clients. Then the other instances connected to Redis will also publish to their clients.
> 
> You don't even need Redis for this, you just need some client and some server to act as "master". I'm pretty sure Kubernetes and the like already has support for master/slave set-ups like this.

Can you explain more in details? I can't imagine how it gonna work.

**Comment by danieleratti** [comment]
It's actually pretty simple:

In your server code you just have to setup a redis PubSub connection or whatever you prefer for message broking...

Then, when you want to send a message to a channel, instead of directly sending it, publish on redis pubsub the channel and the message.

I'll explain better with the following pseudocode:
```
redis = new RedisPubSub()
redis.onMessageReceived(data) {
   ws.publish(data.channel, data.message)
}

// sending a message to a channel
redis.publish({channel: 'foo', message: 'bar'})
```
Instead of the directly publishing on ws.

**Reply by bastienfaivre** [reply]
Hi,
I'm actually trying to implement this in a project. However, I'm facing a problem: since I also need to listen to upcoming message from Redis, I end up with 2 event loops. Indeed, we have the one for uWebSockets and the second one for Redis subscription according to [this example](https://github.com/sewenew/redis-plus-plus#examples-2).
And therefore I don't know how to manage running these two loops simultaneously and properly. How did you manage to do it ?

**Reply by bastienfaivre** [reply]
I managed to do it: https://stackoverflow.com/questions/71896776/using-redis-with-uwebsockets-c-server/71902395#71902395
